botorch/acquisition/joint_entropy_search.py

```diff
@@ -107,5 +285,64 @@ def forward(self, X: Tensor) -> Tensor:
             A `batch_shape`-dim Tensor of acquisition values at the given design
             points `X`.
         """
+        tkwargs = {
+            "dtype": self.optimal_outputs.dtype,
+            "device": self.optimal_outputs.device,
+        }
+
+        initial_posterior = self.initial_model.posterior(X, observation_noise=True)
+
+        add_term = 0.5 * (1 + torch.log(torch.ones(1, **tkwargs) * 2 * pi))
+        batch_shape = X.shape[:-2]
+        sample_dim = len(batch_shape)
+        # We DISREGARD the additional constant term.
+        initial_entropy = 0.5 * torch.logdet(initial_posterior.mvn.covariance_matrix)
+
+        # initial_entropy of shape batch_size or batch_size x num_models if FBGP
+        # first need to unsqueeze the sample dim (first after batch dim) and then the two last
+        initial_entropy = (
+            initial_entropy.unsqueeze(sample_dim).unsqueeze(-1).unsqueeze(-1)
+        )
+
+        CLAMP_LB = torch.finfo(tkwargs["dtype"]).eps
+        # Compute the mixture mean and variance
+        posterior_m = self.conditional_model.posterior(
+            X.unsqueeze(MCMC_DIM), observation_noise=True
+        )
+        noiseless_var = self.conditional_model.posterior(
+            X.unsqueeze(MCMC_DIM), observation_noise=False
+        ).variance
+
+        mean_m = posterior_m.mean
+        if not self.maximize:
+            mean_m = -mean_m
+        variance_m = posterior_m.variance.clamp_min(CLAMP_LB)
+        conditional_samples, conditional_logprobs = self._compute_monte_carlo_variables(
+            posterior_m
+        )
+
+        normalized_samples = (conditional_samples - mean_m) / variance_m.sqrt()
+        # Correlation between noisy observations and noiseless values f
+        rho = (noiseless_var / variance_m).sqrt()
+
+        normal = torch.distributions.Normal(
+            torch.zeros(1, device=X.device, dtype=X.dtype),
+            torch.ones(1, device=X.device, dtype=X.dtype),
+        )
+        # prepare max value quantities and re-scale as required
+        normalized_mvs = (self.optimal_outputs - mean_m) / noiseless_var.sqrt()
+        mvs_rescaled_mc = (normalized_mvs - rho * normalized_samples) / (1 - rho**2)
+        cdf_mvs = normal.cdf(normalized_mvs).clamp_min(CLAMP_LB)
+        cdf_rescaled_mvs = normal.cdf(mvs_rescaled_mc).clamp_min(CLAMP_LB)
+        mv_ratio = cdf_rescaled_mvs / cdf_mvs
+
+        log_term = torch.log(mv_ratio) + conditional_logprobs
```

Comment:
Hmm it would be good to just work with logcdfs from the get go here for numerical stability, but unfortunately I don't think torch has great support for those

Reply:
Looks like my past self noticed that as well: https://github.com/pytorch/pytorch/issues/52973